### PR TITLE
Crp post

### PR DESCRIPTION
**Pull request by ryan-richt**
do it to it

### CODE REVIEW

**Comment by ddgenome** [comment]
It looks like you branched from master instead of develop, so there are a bunch of extra merge commits. I'll have to cherry pick your actual commit into develop once this is approved.

File: _posts/2015-11-17-chinese-restaurant-process.md [hunk 1]
```diff
@@ -0,0 +1,339 @@
...
+
+A multinomial distribution is just like a weighted (unfair) many-sided die, with one side for each outcome. For Christmas M&M's™ say we have a 3-sided die with faces indicating {Red, Green White} which lands on Red 40% of the time, Green 40% of the time and White 20% of the time. To generate a draw of size M (say, 35 candies) from this multinomial you just roll this die M times and count up a vector of each possible color.
+
+![clustering colored candies](/img/MnMs_Clustering.png)
+
+Imagine that the generating function is first buying a bag on a random day. Most of the year you can get bags with classic colors, for 2 weeks you can get Christmas colors, and for 3 days you can get 4th of July colors. So from this distribution of K types of bags say we draw N bags. Then we erase the packaging label of each, and I give you a handful of candies from each bag. Note that there is significant overlap in the colors from each kind of bag. From a small handful of say, Christmas M&M's™ where you didn't chance to draw and white colored candies, its hard to say if you they are Christmas or plain! From only these handfuls, your clustering job is tell me:
```
Comment (ddgenome) [inline _posts/2015-11-17-chinese-restaurant-process.md]: "clustering job is tell me"?

File: _posts/2015-11-17-chinese-restaurant-process.md [hunk 2]
```diff
@@ -0,0 +1,339 @@
...
+
+So this is a clustering process. It is mixture of types of bags, and since we don't get the see the labels of the bags we have a mixture model. The unknown kinds of bags are the clusters, and the handfuls are our data. The multinomial counts of colors from the handfuls are our features extracted from the data, and we describe each type of bag by an explicit probability model which is a multinomial distribution.
+
+Contrast this to say, K-means clustering:
+
+| Attribute         | K-means | Bayesian Mixture Model |
+|-------------------|---------|------------------------|
+| Count of clusters | Ad hoc - user specified "K"  | Probabilistic model |
+| Membership Measure| Ad hoc - euclidean "distance" | Probabilistic model |
+| Solution Method   | Ad hoc - EM-style iterations | Markov-chain Monte Carlo|
```
Comment (ddgenome) [inline _posts/2015-11-17-chinese-restaurant-process.md]: Is EM (expectation maximum) defined anywhere?

File: _posts/2015-11-17-chinese-restaurant-process.md [hunk 3]
```diff
@@ -0,0 +1,339 @@
...
+
+We were then very exciting to be able to run the Gibbs sampler in reasonable memory. Unfortunately we immediately hit the next problem: a single "reseating" of customers in the restaurant took 32.5 seconds. 32.5 seconds x 100,000 customers x 10,000 iterations = _1000 CPU years_. Ouch.
+
+![CRP_cpu_time_reduction.png](/img/CRP_cpu_time_reduction.png)
+
+Using a combination of the sampler and profiler in [VisualVM](https://visualvm.java.net/gettingstarted.html), manual timings, and micro-benchmarks we crafted a series of 7 versions that drove the reseating time down to 0.0001 seconds. Here are some of the highlights:
+
+#### Initialization
+
+If you read papers on CRP, you can see that there are numerous initialization strategies: 1 object per table, log N tables with random assignments, using _Daume 2007_ output as initialization, etc. While 1 object per table seems like the least likely to be biased based on random early decisions, it is
```
Comment (ddgenome) [inline _posts/2015-11-17-chinese-restaurant-process.md]: it is?

**Comment by ddgenome** [comment]
The CRP post looks good to me. @slbahr ?

File: _posts/2015-11-17-chinese-restaurant-process.md (hunk 1 above)
Comment (ryan-richt) [inline _posts/2015-11-17-chinese-restaurant-process.md]: Is to tell me

Sent from my iPhone

On Nov 18, 2015, at 7:32 PM, David Dooling <notifications@github.com<mailto:notifications@github.com>> wrote:


In _posts/2015-11-17-chinese-restaurant-process.md<https://github.com/MonsantoCo/engineering-blog/pull/61#discussion_r45289017>:

> +
> +In Monsanto Labs, we are Bayesians, and as the late [E.T. Jaynes](http://bayes.wustl.edu) espoused, we don't believe in "ad hockeries" like K-means (a numerical method) or ad hoc "machine learning" techniques such as random forests. Instead we have a better way: using only the laws of probability theory. Clustering is actually a difficult problem to cast in the Bayesian paradigm, but new theoretical results and the rise of computing power over the past few decades have made this problem tractable.
> +
> +Proper Bayesian models are "generative," meaning that they posit an underlying (or latent) "generating" process that creates the data we see. It is precisely writing a computer program to recreate the observed data, perhaps with some input variables missing that we want to recover. [Markov-chain Monte Carlo (MCMC)](https://en.wikipedia.org/wiki/Markov_chain_Monte_Carlo) then provides a universal mechanism to "invert" or solve for those input variables for such programs given some data. In the simplest case, say we observe the heights of a room of N people. The generating function could be, a normal distribution with some mean and variance that we draw N samples from. MCMC could then be run on those samples to try to recover the most likely parameters of that normal distribution.
> +
> +We can also construct more complex generating functions (and still solve them with MCMC). Perhaps a better generating function would be to draw Male vs Female from a [binomial distribution](https://en.wikipedia.org/wiki/Binomial_distribution) (probability of being female in the population, like a weighted coin toss), and conditional on the result, draw a height from either the male-specific or female-specific height distribution.
> +
> +### Bayesian Clustering
> +
> +Mixture models are a Bayesian way of clustering: your generating function produces a mixed population of entities from an underlying discrete set of components. For instance, imagine I give you a stream of unlabeled bags of M&M's(tm)? candies. All you get to observe is a few colored M&Ms(tm) of each bag. This is [multinomial](https://en.wikipedia.org/wiki/Multinomial_distribution) count data: we have a finite discrete "vocabulary" of colors and we will observe some number of counts of each color.
> +
> +A multinomial distribution is just like a weighted (unfair) many-sided die, with one side for each outcome. For Christmas M&M's(tm) say we have a 3-sided die with faces indicating {Red, Green White} which lands on Red 40% of the time, Green 40% of the time and White 20% of the time. To generate a draw of size M (say, 35 candies) from this multinomial you just roll this die M times and count up a vector of each possible color.
> +
> +clustering colored candies
> +
> +Imagine that the generating function is first buying a bag on a random day. Most of the year you can get bags with classic colors, for 2 weeks you can get Christmas colors, and for 3 days you can get 4th of July colors. So from this distribution of K types of bags say we draw N bags. Then we erase the packaging label of each, and I give you a handful of candies from each bag. Note that there is significant overlap in the colors from each kind of bag. From a small handful of say, Christmas M&M's(tm) where you didn't chance to draw and white colored candies, its hard to say if you they are Christmas or plain! From only these handfuls, your clustering job is tell me:


"clustering job is tell me"?

-
Reply to this email directly or view it on GitHub<https://github.com/MonsantoCo/engineering-blog/pull/61/files#r45289017>.
This e-mail message may contain privileged and/or confidential information, and is intended to be received only by persons entitled
to receive such information. If you have received this e-mail in error, please notify the sender immediately. Please delete it and
all attachments from any servers, hard drives or any other media. Other use of this e-mail by you is strictly prohibited.

All e-mails and attachments sent and received are subject to monitoring, reading and archival by Monsanto, including its
subsidiaries. The recipient of this e-mail is solely responsible for checking for the presence of "Viruses" or other "Malware".
Monsanto, along with its subsidiaries, accepts no liability for any damage caused by any such code transmitted by or accompanying
this e-mail or any attachment.


The information contained in this email may be subject to the export control laws and regulations of the United States, potentially
including but not limited to the Export Administration Regulations (EAR) and sanctions regulations issued by the U.S. Department of
Treasury, Office of Foreign Asset Controls (OFAC).  As a recipient of this information you are obligated to comply with all
applicable U.S. export laws and regulations.

**Comment by ryan-richt** [comment]
@slbahr tomorrow (monday) is my last day girl, can you get to this in the morning?

**Comment by slbahr** [comment]
Looks good!  Go for it!

**Comment by ddgenome** [comment]
I am going to close this PR and merge in the single post manually.

**Pull request closed without merging**